coreos-virt-install

```diff
@@ -0,0 +1,92 @@
+    """Determine the number of CPUs that should be visible in the guest.
+    See e.g. https://www.redhat.com/archives/libvirt-users/2017-June/msg00025.html
+    We want to match the number of host physical cores.
+    """
+    # We may be run in a cgroup with fewer cores available than physical.
+    available_cpu = int(subprocess.check_output(['nproc']).strip())
+    # https://stackoverflow.com/questions/6481005/how-to-obtain-the-number-of-cpus-cores-in-linux-from-the-command-line
+    core_sockets = set()
+    for line in io.BytesIO(subprocess.check_output(['lscpu', '-b', '-p=Core,Socket'])):
+        if line.startswith(b'#'):
+            continue
+        core_sockets.add(line.strip())
+    sockets = min(available_cpu, len(core_sockets))
+    return '--vcpus=sockets={},cores=1,threads=1'.format(sockets, sockets)
+
+def run_sync_verbose(argv, **kwargs):
+    print("cmd: {}".format(subprocess.list2cmdline(argv)))
+    subprocess.check_call(argv, **kwargs)
+
+# This is an "extension" to kickstart files that we implement as a comment.
+if args.create_disk:
+    magic_virt_install_size_str = '#--virt-install-disk-size-gb:'
+    disk_re = re.compile(r'^' + magic_virt_install_size_str + ' *([0-9]+)$')
+    disk_size = None
+    with open(args.kickstart) as f:
+        for line in f:
+            m = disk_re.search(line)
+            if not m:
+                continue
+            disk_size = int(m.group(1))
+            break
+    if disk_size is None:
+        fatal("--create-disk specified, but failed to find '{}' in kickstart".format(magic_virt_install_size_str))
+    run_sync_verbose(['qemu-img', 'create', '-f', 'qcow2', args.dest, '{}G'.format(disk_size)])
+    print("Created initial disk: {} with size {}G".format(args.dest, disk_size))
+
+# Doesn't need to be truly unique, let's just use our pid+time
+domain="coreos-inst-{}-{}".format(os.getpid(),int(time.time()))
+try:
+    vinstall_args = ["virt-install", "--connect=qemu:///session",
+                     "--name={}".format(domain),
+                     "--noautoconsole"]
+    if args.console_log_file:
+        vinstall_args.append("--console=log.file={}".format(args.console_log_file))
+    vinstall_args.extend(["--wait={}".format(args.wait), "--noreboot", "--nographics",
+                          "--memory={}".format(args.memory), get_libvirt_smp_arg(),
+                          "--os-variant=rhel7", "--rng=/dev/urandom",
+                          "--check", "path_in_use=off",
+                          "--disk=path={},cache=unsafe".format(args.dest),
+                          "--location={}".format(args.location),
+                          "--initrd-inject={}".format(args.kickstart),
+                          "--extra-args", "ks=file://{} console=tty0 console=ttyS0,115200n8 inst.cmdline inst.notmux".format(os.path.basename(args.kickstart))])
+    run_sync_verbose(vinstall_args)
+finally:
+    subprocess.call(['virsh', '--connect=qemu:///session', 'undefine', domain])
```

Comment:
I see `virt-install` does support a `--transient` option here. It doesn't *seem* like it would work in our case from the description, though it'd be really neat if it did. Did you experiment with it?

Reply:
Yep I tried, it's actually not what we want - it creates a snapshot of the disk, whereas we obviously want the contents of the disk to live on after qemu is gone.

Reply:
Ahh, that's too bad.

Reply:
Though in practice in prod since we're running this in a disposable container, it's all cleaned up automatically by that, so this is just a "nice to have" for the pet container case.

Reply:
(This is all a notable advancement over e.g. Koji+ImageFactory today where "leaked" libvirt images are a real problem)